Source/Core/Core/HW/GCMemcard/GCMemcard.cpp

```diff
@@ -27,183 +27,181 @@ static void ByteSwap(u8* valueA, u8* valueB)
+         Test(GCMemcardValidityIssues::INVALID_CHECKSUM) ||
+         Test(GCMemcardValidityIssues::FREE_BLOCK_MISMATCH);
+}
 
-    m_size_mb = (u16)((size / BLOCK_SIZE) / MBIT_TO_BLOCKS);
-    switch (m_size_mb)
-    {
-    case MemCard59Mb:
-    case MemCard123Mb:
-    case MemCard251Mb:
-    case Memcard507Mb:
-    case MemCard1019Mb:
-    case MemCard2043Mb:
-      break;
-    default:
-      PanicAlertT("%s failed to load as a memory card.\nCard size is invalid (0x%x bytes)",
-                  filename.c_str(), (unsigned)size);
-      return;
-    }
-  }
+bool GCMemcardErrorCode::Test(GCMemcardValidityIssues code) const
+{
+  return m_errors.test(static_cast<size_t>(code));
+}
 
-  mcdFile.Seek(0, SEEK_SET);
-  if (!mcdFile.ReadBytes(&m_header_block, BLOCK_SIZE))
-  {
-    PanicAlertT("Failed to read header correctly\n(0x0000-0x1FFF)");
-    return;
-  }
-  if (m_size_mb != m_header_block.m_size_mb)
-  {
-    PanicAlertT("Memory card file size does not match the header size");
-    return;
-  }
+void GCMemcardErrorCode::Set(GCMemcardValidityIssues code)
+{
+  m_errors.set(static_cast<size_t>(code));
+}
 
-  if (!mcdFile.ReadBytes(&m_directory_blocks[0], BLOCK_SIZE))
-  {
-    PanicAlertT("Failed to read 1st directory block correctly\n(0x2000-0x3FFF)");
-    return;
-  }
+GCMemcardErrorCode& GCMemcardErrorCode::operator|=(const GCMemcardErrorCode& other)
+{
+  this->m_errors |= other.m_errors;
+  return *this;
+}
 
-  if (!mcdFile.ReadBytes(&m_directory_blocks[1], BLOCK_SIZE))
-  {
-    PanicAlertT("Failed to read 2nd directory block correctly\n(0x4000-0x5FFF)");
-    return;
-  }
+GCMemcard::GCMemcard()
+    : m_valid(false), m_size_blocks(0), m_size_mb(0), m_active_directory(0), m_active_bat(0)
+{
+}
+
+std::optional<GCMemcard> GCMemcard::Create(std::string filename, u16 size_mbits, bool shift_jis)
+{
+  GCMemcard card;
+  card.m_filename = std::move(filename);
 
-  if (!mcdFile.ReadBytes(&m_bat_blocks[0], BLOCK_SIZE))
+  // TODO: Format() not only formats the card but also writes it to disk at m_filename.
+  // Those tasks should probably be separated.
+  if (!card.Format(shift_jis, size_mbits))
+    return std::nullopt;
+
+  return std::move(card);
+}
+
+std::pair<GCMemcardErrorCode, std::optional<GCMemcard>> GCMemcard::Open(std::string filename)
+{
+  GCMemcardErrorCode error_code;
+  File::IOFile file(filename, "rb");
+  if (!file.IsOpen())
   {
-    PanicAlertT("Failed to read 1st block allocation table block correctly\n(0x6000-0x7FFF)");
-    return;
+    error_code.Set(GCMemcardValidityIssues::FAILED_TO_OPEN);
+    return std::make_pair(error_code, std::nullopt);
   }
 
-  if (!mcdFile.ReadBytes(&m_bat_blocks[1], BLOCK_SIZE))
+  // check if the filesize is a valid memory card size
+  u16 card_size_mbits;
+  u64 filesize = file.GetSize();
+  switch (filesize)
   {
-    PanicAlertT("Failed to read 2nd block allocation table block correctly\n(0x8000-0x9FFF)");
-    return;
+  case (MemCard59Mb * MBIT_TO_BLOCKS) * BLOCK_SIZE:
```

Comment:
What do you think of a small constexpr function to generate this math?

Example:
```suggestion
// Somewhere else:
constexpr u64 MemCardSizeToFileSize(u16 card_size) {
  return card_size * MBIT_TO_BLOCKS * BLOCK_SIZE;
}

// case label:
  case MemCardSizeToFileSize(MemCard59Mb):
```

Reply:
I made a more generic `MegabitsToBytes()` instead but yeah, that's reasonable.